src/subscriber.ts

```diff
@@ -14,497 +14,351 @@
+ */
+export class Subscriber extends EventEmitter {
+  ackDeadline: number;
+  isOpen: boolean;
+  private _acks!: AckQueue;
+  private _client!: ClientStub;
+  private _histogram: Histogram;
+  private _inventory!: LeaseManager;
+  private _isUserSetDeadline: boolean;
+  private _latencies: Histogram;
+  private _modAcks!: ModAckQueue;
+  private _name!: string;
+  private _options!: SubscriberOptions;
+  private _stream!: MessageStream;
+  private _subscription: Subscription;
+  constructor(subscription: Subscription, options = {}) {
+    super();
 
-      Promise.all(modAckRequests).then(() => {
-        this.inventory_.nack = [];
-      });
-    }
+    this.ackDeadline = 10;
+    this.isOpen = false;
+    this._isUserSetDeadline = false;
+    this._histogram = new Histogram({min: 10, max: 600});
+    this._latencies = new Histogram();
+    this._subscription = subscription;
 
-    return Promise.all(requests);
+    this.setOptions(options);
   }
-  /*!
-   * Checks to see if we currently have a streaming connection.
-   *
-   * @private
+  /**
+   * The 99th percentile of request latencies.
    *
-   * @return {boolean}
+   * @type {number}
    */
-  isConnected_() {
-    return !!(this.connectionPool && this.connectionPool.isConnected());
+  get latency() {
+    return this._latencies.percentile(99);
   }
-  /*!
-   * Checks to see if this Subscriber has hit any of the flow control
-   * thresholds.
-   *
-   * @private
+  /**
+   * The full name of the Subscription.
    *
-   * @return {boolean}
+   * @type {string}
    */
-  hasMaxMessages_() {
-    return (
-        this.inventory_.lease.length >= this.flowControl.maxMessages ||
-        this.inventory_.bytes >= this.flowControl.maxBytes);
+  get name(): string {
+    if (!this._name) {
+      const {name, projectId} = this._subscription;
+      this._name = replaceProjectIdToken(name, projectId);
+    }
+
+    return this._name;
   }
-  /*!
-   * Leases a message. This will add the message to our inventory list and then
-   * modifiy the ack deadline for the user if they exceed the specified ack
-   * deadline.
-   *
-   * @private
+  /**
+   * Acknowledges the supplied message.
    *
-   * @param {object} message The message object.
+   * @param {Message} message The message to acknowledge.
+   * @returns {Promise}
    */
-  leaseMessage_(message) {
-    this.modifyAckDeadline_(
-        message.ackId, this.ackDeadline / 1000, message.connectionId);
-    this.inventory_.lease.push(message.ackId);
-    this.inventory_.bytes += message.length;
-    this.setLeaseTimeout_();
-    return message;
+  async ack(message: Message): Promise<void> {
+    const startTime = Date.now();
+
+    if (!this._isUserSetDeadline) {
+      const ackTimeSeconds = (startTime - message.received) / 1000;
+      this._histogram.add(ackTimeSeconds);
+      this.ackDeadline = this._histogram.percentile(99);
+    }
+
+    this._acks.add(message);
+    await this._acks.onFlush();
```

Comment:
Not sure I understand, `onFlush` is an event we are tapping into. Since acks/modAcks get queued up, we're basically saying hey I queued this message, let me know once it gets flushed from the queue.